catboost/libs/metrics/metric.cpp

```diff
@@ -1779,6 +1779,128 @@ TVector<TParamSet> TTweedieMetric::ValidParamSets() {
+        explicit TFocalMetric(const TLossParams& params, double focal_alpha, double focal_gamma)
+            : TAdditiveMetric(ELossFunction::Focal, params)
+            , FocalAlpha(focal_alpha), FocalGamma(focal_gamma) {
+            CB_ENSURE(FocalAlpha > 0 && FocalAlpha < 1, "Focal metric is defined for 0 < focal_alpha < 1, got " << focal_alpha);
+            CB_ENSURE(FocalGamma > 0, "Focal metric is defined for 0 < focal_gamma, got " << focal_gamma);
+        }
+
+        static TVector<THolder<IMetric>> Create(const TMetricConfig& config);
+        static TVector<TParamSet> ValidParamSets();
+
+        TMetricHolder EvalSingleThread(
+                const TConstArrayRef<TConstArrayRef<double>> approx,
+                const TConstArrayRef<TConstArrayRef<double>> approxDelta,
+                bool isExpApprox,
+                TConstArrayRef<float> target,
+                TConstArrayRef<float> weight,
+                TConstArrayRef<TQueryInfo> queriesInfo,
+                int begin,
+                int end
+        ) const override;
+        void GetBestValue(EMetricBestValue* valueType, float* bestValue) const override;
+
+    private:
+        const double FocalAlpha;
+        const double FocalGamma;
+    };
+}
+
+// static
+TVector<THolder<IMetric>> TFocalMetric::Create(const TMetricConfig& config) {
+    CB_ENSURE(config.GetParamsMap().contains("focal_alpha"), "Metric " << ELossFunction::Focal << " requires focal_alpha as parameter");
+    CB_ENSURE(config.GetParamsMap().contains("focal_gamma"), "Metric " << ELossFunction::Focal << " requires focal_gamma as parameter");
+    config.ValidParams->insert("focal_alpha");
+    config.ValidParams->insert("focal_gamma");
+    return AsVector(MakeHolder<TFocalMetric>(config.Params,
+                                               FromString<float>(config.GetParamsMap().at("focal_alpha")),
+                                               FromString<float>(config.GetParamsMap().at("focal_gamma"))
+                                            )
+                    );
+}
+
+TMetricHolder TFocalMetric::EvalSingleThread(
+        const TConstArrayRef<TConstArrayRef<double>> approx,
+        const TConstArrayRef<TConstArrayRef<double>> approxDelta,
+        bool isExpApprox,
+        TConstArrayRef<float> target,
+        TConstArrayRef<float> weight,
+        TConstArrayRef<TQueryInfo> /*queriesInfo*/,
+        int begin,
+        int end
+) const {
+    CB_ENSURE(approx.size() == 1, "Metric Focal supports only single-dimensional data");
+    Y_ASSERT(!isExpApprox);
+    const auto impl = [=] (auto hasDelta, auto hasWeight, TConstArrayRef<double> approx, TConstArrayRef<double> approxDelta) {
+        TMetricHolder error(2);
+        for (int k : xrange(begin, end)) {
+            double curApprox = approx[k];
+            if (hasDelta) {
+                curApprox += approxDelta[k];
+            }
+            const float w = hasWeight ? weight[k] : 1;
+            double FocalAlpha = 0.75;
```

Comment:
It is necessary to pass these parameters into the metric somehow.

Reply:
I'd better wait for maintainers' advice on how to do it in the most correct way(probably in the way I did it with loss params)